sqlite/sqlite.go

```diff
@@ -230,14 +231,38 @@ func (s *SqlStore) RestoreSqlStore(ctx context.Context, r io.Reader) error {
 		return err
 	}
 
-	// Close the current DB.
+	// Close the current DB and run the restore while under lock.
+	s.Mu.Lock()
+	if err := s.runSqlRestore(ctx, tempFileName); err != nil {
+		s.Mu.Unlock()
+		return err
+	}
+
+	// Now that the DB has been re-opened, release the lock so that the migrations
+	// can be run.
+	s.Mu.Unlock()
+
+	// Run migrations on the restored database. This will catch the restored
+	// database up to the current state expected by the application.
+	restoreMigrator := NewMigrator(
+		s,
+		s.log.With(zap.String("service", "sqlite restore migrations")),
+	)
+
+	return restoreMigrator.Up(ctx, sqliteMigrations.All)
+}
+
+// swapAndOpenRestored is used while restoring a database to close the existing
+// database, restore the backed up database, and then re-open the restored
+// database.
+func (s *SqlStore) runSqlRestore(ctx context.Context, tempFileName string) error {
```

Comment:
As above in the bolt kv restore file, this function was adapted from existing code to make it easier to manage the lock.